### PR TITLE
Reduce unnecessary whitespace in JSON

### PR DESCRIPTION
**Pull request by auscompgeek**
The `json` and `simplejson` modules when serialising default to separators such that the output is compatible with YAML 1.0.  As there is no need for JSON responses from Matrix homeservers to be consumed by YAML 1.0/1.1 implementations, the whitespace this generates is redundant.

This PR reduces the amount of redundant whitespace that is stored and sent in responses, hopefully significantly reducing the resources required for Synapse in the long run.

This does not attempt to eliminate the unnecessary whitespace completely, nor does it introduce anything to rewrite data in existing databases to be compact, but rather this tries to find the biggest future wins for the average homeserver.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog).
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

`Signed-off-by: David Vo <david@vovo.id.au>`

### CODE REVIEW

**Comment by richvdh** [comment]
my main concern here would be if it causes us to fall back to the pure-python implementation of json.dumps instead of the optimised C version.

**Comment by auscompgeek** [comment]
Both `json` and `simplejson` look like they both have non-default separators support in their C extension modules.

Taking a look though, it seems `dumps` internally creates a new `JSONEncoder` instance if passed non-default options though. Probably worth creating our own `JSONEncoder` instance and calling `encode` on it?

**Comment by auscompgeek** [comment]
Anyway, if falling back to the pure-Python implementation was a concern here, I would have some bad news about `canonicaljson`...

**Comment by richvdh** [comment]
> Probably worth creating our own JSONEncoder instance and calling encode on it?

yes, that sounds sensible.

> Anyway, if falling back to the pure-Python implementation was a concern here, I would have some bad news about `canonicaljson`...

it is a concern, so please share your bad news. canonicaljson goes to [some lengths](https://github.com/matrix-org/python-canonicaljson/blob/master/canonicaljson.py#L47-L49) to use the C implementation.

**Comment by auscompgeek** [comment]
> > Probably worth creating our own JSONEncoder instance and calling encode on it?
> 
> yes, that sounds sensible.

What would be a good place for this to go? Seems like it could go in `synapse.util` but I can't see a good existing module to put it in, and it seems like overkill to create an entire new module for a single object.

> canonicaljson goes to [some lengths](https://github.com/matrix-org/python-canonicaljson/blob/master/canonicaljson.py#L47-L49) to use the C implementation.

That is an interesting interpretation of that comment, given that [both paths](https://github.com/python/cpython/blob/v3.8.2/Lib/json/encoder.py#L218) are [implemented](https://github.com/python/cpython/blob/v3.8.2/Lib/json/encoder.py#L45) in C.

**Comment by richvdh** [comment]
>  What would be a good place for this to go? 

I guess it could just go in `synapse/util/__init__.py` :man_shrugging:

**Comment by richvdh** [comment]
> > canonicaljson goes to [some lengths](https://github.com/matrix-org/python-canonicaljson/blob/master/canonicaljson.py#L47-L49) to use the C implementation.
> 
> That is an interesting interpretation of that comment, 

"that comment" being "if falling back to the pure-Python implementation was a concern here, I would have some bad news about canonicaljson..."?

My interpretation of that was that you believe that canonicaljson is falling back to pure-python. Looking again, it seems that my interpretation of your comment was wrong? Either way, I'd appreciate it if you could be clear rather than making me guess at your meaning.

> given that [both paths](https://github.com/python/cpython/blob/v3.8.2/Lib/json/encoder.py#L218) are [implemented](https://github.com/python/cpython/blob/v3.8.2/Lib/json/encoder.py#L45) in C.

again, these links don't help me without spending 15 minutes staring at the context. Please: be clear what I'm supposed to understand here.

**Comment by auscompgeek** [comment]
I would, however I won't have an opportunity to work on this until next month. Happy for anyone to take over.

**Comment by richvdh** [comment]
Thanks, we'll wait! :)

**Comment by auscompgeek** [comment]
I suppose 0364266eb734cb5b99be3a8afc4ddcd6f32bf3a1 could be merged separately though. That should give the biggest storage wins.

**Comment by clokep** [comment]
> I suppose [0364266](https://github.com/matrix-org/synapse/commit/0364266eb734cb5b99be3a8afc4ddcd6f32bf3a1) could be merged separately though. That should give the biggest storage wins.

@auscompgeek That makes sense to me, I don't think that would be a controversial change either.

**Comment by auscompgeek** [comment]
Okay, I've rewritten this PR to directly use a `JSONEncoder`.

**Comment by clokep** [comment]
@auscompgeek Would you mind merging develop into this? I think with some of the changes made for #7674 it would be nice to have that in here for review to ensure there's no issues with compatibility.

File: synapse/util/__init__.py [hunk 1]
```diff
@@ -17,13 +17,16 @@
...
 import attr
+from canonicaljson import json
 
 from twisted.internet import defer, task
 
 from synapse.logging import context
 
 logger = logging.getLogger(__name__)
 
+json_encoder = json.JSONEncoder(separators=(",", ":"))
```
Comment (richvdh) [inline synapse/util/__init__.py:28]: please could you give this a docstring? where is it intended that it be used?

Comment (richvdh) [reply synapse/util/__init__.py:28]: oh, docstrings don't really work on variables like this. A comment instead?

Comment (clokep) [reply synapse/util/__init__.py:28]:
```suggestion
# Create a custom encoder to reduce the whitespace produced by JSON encoding.
json_encoder = json.JSONEncoder(separators=(",", ":"))
```

File: synapse/util/frozenutils.py [hunk 2]
```diff
@@ -64,4 +64,6 @@ def _handle_frozendict(obj):
 
 
 # A JSONEncoder which is capable of encoding frozendicts without barfing
```
Comment (clokep) [inline synapse/util/frozenutils.py:66]:
```suggestion
# A JSONEncoder which is capable of encoding frozendicts without barfing.
# Additionally reduce the whitespace produced by JSON encoding.
```

**Comment by clokep** [comment]
@auscompgeek I left a couple of suggestions to improve the comments. Additionally it seems that this bitrotted. Hopefully git will mostly do the right thing when merging in develop! I think this looks good and should be a great change!

**Comment by clokep** [comment]
It looks like this could also replace the custom encoder in [`synapse/synapse/replication/tcp/commands.py`](https://github.com/matrix-org/synapse/blob/f299441cc67f31dcd47b8fdfda4a218bee9df9ba/synapse/replication/tcp/commands.py#L25), no reason not to reduce the traffic sent over replication!

**Comment by auscompgeek** [comment]
Thanks for the comment suggestions :)

> Additionally it seems that this bitrotted. Hopefully git will mostly do the right thing when merging in develop!

Yup, this PR seems to have a bit of a habit of bitrotting, thank goodness for `git config merge.conflictstyle diff3` though.

> replace the custom encoder in [`synapse/replication/tcp/commands.py`](https://github.com/matrix-org/synapse/blob/f299441cc67f31dcd47b8fdfda4a218bee9df9ba/synapse/replication/tcp/commands.py#L25)

Good catch. It looks like this uses the stdlib `json` directly though, do we want to change that?

Looking around, I found another `JSONEncoder` as well in `synapse.logging._terse_json`:

https://github.com/matrix-org/synapse/blob/2ffd6783c7af12e3c29e1a44dee4a9deeb83890b/synapse/logging/_terse_json.py#L42

That one uses both `ensure_ascii=False` and the stdlib `json` directly, so I won't touch that one here.

**Comment by clokep** [comment]
> > replace the custom encoder in [`synapse/replication/tcp/commands.py`](https://github.com/matrix-org/synapse/blob/f299441cc67f31dcd47b8fdfda4a218bee9df9ba/synapse/replication/tcp/commands.py#L25)
> 
> Good catch. It looks like this uses the stdlib `json` directly though, do we want to change that?

I think changing this one will be fine.

I agree that leaving the `synapse.logging._terse_json` one alone makes sense.

**Comment by clokep** [comment]
Thanks for sticking with us on this one! 👍

**Pull request merged**